cra-rxjs-styled-components/src/hooks/gists/use-gists.ts

```diff
@@ -0,0 +1,87 @@
+import { useEffect, useState } from 'react';
+import { GITHUB_URL_BASE, GISTS_URL } from '../../constants/url.constants';
+import { Gist, GistWithFilename } from '../../interfaces/gists.interfaces';
+import { fromFetchWithAuth } from '../auth/from-fetch-with-auth';
+import {
+  filter,
+  map,
+  Subscription,
+  Observable,
+  switchMap,
+  tap,
+  zip,
+} from 'rxjs';
+
+export function useGists(): GistWithFilename[] {
+  const [state, setState] = useState<GistWithFilename[]>([]);
+  const [currentUser, setCurrentUser] = useState<string>('');
+
+  const getCurrentUser = () =>
+    fromFetchWithAuth<{
+      login: string;
+    }>(`${GITHUB_URL_BASE}/user`, {
+      selector: (response: Response) => {
+        return response.json();
+      },
+    });
+
+  useEffect(() => {
+    getCurrentUser().subscribe((user) => {
+      setCurrentUser(user.login);
+    });
+
+    const subscription: Subscription = fromFetchWithAuth<Gist[]>(
+      GISTS_URL(currentUser),
+      {
+        selector: (response: Response) => {
+          return response.json();
+        },
+      }
+    )
+      .pipe(
+        filter((gists) => !!gists.length),
+        switchMap((gists: Gist[]) => {
+          const requests = gists.map(createGistRequest);
+
+          return zip(...requests).pipe(
+            map((gists) => {
+              return mergeGists(gists);
+            })
+          );
+        }),
+        tap(setState)
+      )
+      .subscribe();
+
+    return () => {
+      subscription.unsubscribe();
+    };
+  }, [currentUser]);
+
+  return state;
+}
+
+function createGistRequest(gist: Gist): Observable<Gist> {
+  return fromFetchWithAuth<Gist>(gist.url, {
+    selector: (response: Response) => {
+      return response.json();
+    },
+  });
+}
```

Comment:
Is there not a way to just fetch a list of the most recent gists with the data needed on them? It's been a while since I did one of these but I don't remember having to do it like this. Seems like too much work going on here just to get and parse a small list of items

Reply:
I'm not sure what you mean with this? This is what I'm doing with this function.

Reply:
It looks like it makes a request to get a list of gists, and then creates individual requests to fetch each gist based on that